src/caffe/layers/cudnn_conv_layer.cu

```diff
@@ -21,21 +21,54 @@ void CuDNNConvolutionLayer<Dtype>::Forward_gpu(
 
     // Forward through cuDNN in parallel over groups.
     for (int g = 0; g < this->group_; g++) {
+      Dtype alpha = 1.0;
+      Dtype beta = 0.0;
+
+      cudnnConvolutionFwdAlgo_t algo;
+
+      // get the desired convolution algorithm
+      CUDNN_CHECK(cudnnGetConvolutionForwardAlgorithm(handle_[g],
+        bottom_descs_[i],
+        filter_desc_,
+        conv_descs_[i],
+        top_descs_[i],
+        CUDNN_CONVOLUTION_FWD_NO_WORKSPACE,
+        0,  // memoryLimitInBytes,
+        &algo));
+
+      // get minimum size of the workspace needed for the desired algorithm
+      size_t workspaceSizeInBytes;
+
+      CUDNN_CHECK(cudnnGetConvolutionForwardWorkspaceSize(handle_[g],
+        bottom_descs_[i],
+        filter_desc_,
+        conv_descs_[i],
+        top_descs_[i],
+        algo,
+        &workspaceSizeInBytes));
+
+      void *workspace = NULL;
+
       // Filters.
       CUDNN_CHECK(cudnnConvolutionForward(handle_[g],
-          bottom_descs_[i], bottom_data + bottom_offset_ * g,
-          filter_desc_, weight + weight_offset_ * g,
-          conv_descs_[i],
-          top_descs_[i], top_data + top_offset_ * g,
-          CUDNN_RESULT_NO_ACCUMULATE));
+            reinterpret_cast<void *>(&alpha),
```

Comment:
Right, C-style casting is right out; whether one should use `reinterpret_cast` or `static_cast` here is a stylistic issue that can be argued either way. What bothers me is that the `alpha` and `beta` arguments are being treated differently than the data arguments, even though (according to the header I am looking at), all of those arguments have type `void*` and are subject to the same danger. Is there a reason for that?